<?xml version="1.0"?>

<!-- Robot Name --> 
<robot name="LiCAS_A1" xmlns:xacro="http://ros.org/wiki/xacro">

<!-- Import Transmissions -->
<xacro:include filename="$(find uav-cable-suspended-robots-ros-pkg)/urdf/licasa1.transmission.xacro" />
<xacro:LiCAS_A1_transmission name="LiCAS_A1" hardware_interface1="hardware_interface/PositionJointInterface" hardware_interface2="hardware_interface/EffortJointInterface" /> 

<!-- Import Gazebo plugins -->
<xacro:include filename="$(find uav-cable-suspended-robots-ros-pkg)/gazebo/gazebo_licasa1.urdf.xacro"/> 

<link name="base_link"/>
  <joint name="base_joint" type="fixed">
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <parent link="base_link"/>
    <child link="base_link_inertia"/>
  </joint>
  <link name="base_link_inertia">
    <inertial>
      <mass value="1.5"/>
      <!-- [kg] -->
      <origin xyz="0 0 0"/>
      <inertia ixx="0.0347563" ixy="0.0" ixz="0.0" iyy="0.0458929" iyz="0.0" izz="0.0977"/>
    </inertial>
    <visual>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <mesh filename="package://rotors_description/meshes/firefly.dae" scale="1 1 1"/>
      </geometry>
    </visual>
    <collision>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <geometry>
        <box size="0.1 0.1 0.16"/>
        <!-- [m] [m] [m] -->
      </geometry>
    </collision>
  </link>
  <!-- attach multirotor_base_plugin to the base_link -->
  <gazebo>
    <plugin filename="librotors_gazebo_multirotor_base_plugin.so" name="multirotor_base_plugin">
      <robotNamespace>licasa1</robotNamespace>
      <linkName>base_link</linkName>
      <rotorVelocitySlowdownSim>10</rotorVelocitySlowdownSim>
    </plugin>
  </gazebo>
  <joint name="rotor_0_joint" type="continuous">
    <origin rpy="0 0 0" xyz="0.18619546181355998 0.1075 0.037"/>
    <axis xyz="0 0 1"/>
    <!-- TODO(ff): not currently set because it's not yet supported -->
    <!-- <limit effort="2000" velocity="${max_rot_velocity}" /> -->
    <parent link="base_link"/>
    <child link="rotor_0"/>
  </joint>
  <link name="rotor_0">
    <inertial>
      <mass value="0.005"/>
      <!-- [kg] -->
      <inertia ixx="9.7499961e-07" ixy="0.0" ixz="0.0" iyy="4.170414998500001e-05" iyz="0.0" izz="4.2604149625000006e-05"/>
    </inertial>
    <visual>
      <geometry>
        <mesh filename="package://rotors_description/meshes/propeller_ccw.dae" scale="0.1 0.1 0.1"/>
        <!-- The propeller meshes have a radius of 1m -->
        <!-- <box size="${2*radius_rotor} 0.01 0.005"/> -->
      </geometry>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.005" radius="0.1"/>
        <!-- [m] -->
      </geometry>
    </collision>
  </link>
  <gazebo>
    <plugin filename="librotors_gazebo_motor_model.so" name="licasa1_front_left_motor_model">
      <robotNamespace>licasa1</robotNamespace>
      <jointName>rotor_0_joint</jointName>
      <linkName>rotor_0</linkName>
      <turningDirection>ccw</turningDirection>
      <timeConstantUp>0.0125</timeConstantUp>
      <timeConstantDown>0.025</timeConstantDown>
      <maxRotVelocity>838000</maxRotVelocity>
      <motorConstant>8.54858e-06</motorConstant>
      <momentConstant>0.016</momentConstant>
      <commandSubTopic>gazebo/command/motor_speed</commandSubTopic>
      <motorNumber>0</motorNumber>
      <rotorDragCoefficient>8.06428e-05</rotorDragCoefficient>
      <rollingMomentCoefficient>1e-06</rollingMomentCoefficient>
      <motorSpeedPubTopic>motor_speed/0</motorSpeedPubTopic>
      <rotorVelocitySlowdownSim>10</rotorVelocitySlowdownSim>
    </plugin>
  </gazebo>
  <gazebo reference="rotor_0">
    <material>Gazebo/Red</material>
  </gazebo>
  <joint name="rotor_1_joint" type="continuous">
    <origin rpy="0 0 0" xyz="0 0.215 0.037"/>
    <axis xyz="0 0 1"/>
    <!-- TODO(ff): not currently set because it's not yet supported -->
    <!-- <limit effort="2000" velocity="${max_rot_velocity}" /> -->
    <parent link="base_link"/>
    <child link="rotor_1"/>
  </joint>
  <link name="rotor_1">
    <inertial>
      <mass value="0.005"/>
      <!-- [kg] -->
      <inertia ixx="9.7499961e-07" ixy="0.0" ixz="0.0" iyy="4.170414998500001e-05" iyz="0.0" izz="4.2604149625000006e-05"/>
    </inertial>
    <visual>
      <geometry>
        <mesh filename="package://rotors_description/meshes/propeller_cw.dae" scale="0.1 0.1 0.1"/>
        <!-- The propeller meshes have a radius of 1m -->
        <!-- <box size="${2*radius_rotor} 0.01 0.005"/> -->
      </geometry>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.005" radius="0.1"/>
        <!-- [m] -->
      </geometry>
    </collision>
  </link>
  <gazebo>
    <plugin filename="librotors_gazebo_motor_model.so" name="licasa1_left_motor_model">
      <robotNamespace>licasa1</robotNamespace>
      <jointName>rotor_1_joint</jointName>
      <linkName>rotor_1</linkName>
      <turningDirection>cw</turningDirection>
      <timeConstantUp>0.0125</timeConstantUp>
      <timeConstantDown>0.025</timeConstantDown>
      <maxRotVelocity>838000</maxRotVelocity>
      <motorConstant>8.54858e-06</motorConstant>
      <momentConstant>0.016</momentConstant>
      <commandSubTopic>gazebo/command/motor_speed</commandSubTopic>
      <motorNumber>1</motorNumber>
      <rotorDragCoefficient>8.06428e-05</rotorDragCoefficient>
      <rollingMomentCoefficient>1e-06</rollingMomentCoefficient>
      <motorSpeedPubTopic>motor_speed/1</motorSpeedPubTopic>
      <rotorVelocitySlowdownSim>10</rotorVelocitySlowdownSim>
    </plugin>
  </gazebo>
  <gazebo reference="rotor_1">
    <material>Gazebo/Blue</material>
  </gazebo>
  <joint name="rotor_2_joint" type="continuous">
    <origin rpy="0 0 0" xyz="-0.18619546181355998 0.1075 0.037"/>
    <axis xyz="0 0 1"/>
    <!-- TODO(ff): not currently set because it's not yet supported -->
    <!-- <limit effort="2000" velocity="${max_rot_velocity}" /> -->
    <parent link="base_link"/>
    <child link="rotor_2"/>
  </joint>
  <link name="rotor_2">
    <inertial>
      <mass value="0.005"/>
      <!-- [kg] -->
      <inertia ixx="9.7499961e-07" ixy="0.0" ixz="0.0" iyy="4.170414998500001e-05" iyz="0.0" izz="4.2604149625000006e-05"/>
    </inertial>
    <visual>
      <geometry>
        <mesh filename="package://rotors_description/meshes/propeller_ccw.dae" scale="0.1 0.1 0.1"/>
        <!-- The propeller meshes have a radius of 1m -->
        <!-- <box size="${2*radius_rotor} 0.01 0.005"/> -->
      </geometry>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.005" radius="0.1"/>
        <!-- [m] -->
      </geometry>
    </collision>
  </link>
  <gazebo>
    <plugin filename="librotors_gazebo_motor_model.so" name="licasa1_back_left_motor_model">
      <robotNamespace>licasa1</robotNamespace>
      <jointName>rotor_2_joint</jointName>
      <linkName>rotor_2</linkName>
      <turningDirection>ccw</turningDirection>
      <timeConstantUp>0.0125</timeConstantUp>
      <timeConstantDown>0.025</timeConstantDown>
      <maxRotVelocity>838000</maxRotVelocity>
      <motorConstant>8.54858e-06</motorConstant>
      <momentConstant>0.016</momentConstant>
      <commandSubTopic>gazebo/command/motor_speed</commandSubTopic>
      <motorNumber>2</motorNumber>
      <rotorDragCoefficient>8.06428e-05</rotorDragCoefficient>
      <rollingMomentCoefficient>1e-06</rollingMomentCoefficient>
      <motorSpeedPubTopic>motor_speed/2</motorSpeedPubTopic>
      <rotorVelocitySlowdownSim>10</rotorVelocitySlowdownSim>
    </plugin>
  </gazebo>
  <gazebo reference="rotor_2">
    <material>Gazebo/Blue</material>
  </gazebo>
  <joint name="rotor_3_joint" type="continuous">
    <origin rpy="0 0 0" xyz="-0.18619546181355998 -0.1075 0.037"/>
    <axis xyz="0 0 1"/>
    <!-- TODO(ff): not currently set because it's not yet supported -->
    <!-- <limit effort="2000" velocity="${max_rot_velocity}" /> -->
    <parent link="base_link"/>
    <child link="rotor_3"/>
  </joint>
  <link name="rotor_3">
    <inertial>
      <mass value="0.005"/>
      <!-- [kg] -->
      <inertia ixx="9.7499961e-07" ixy="0.0" ixz="0.0" iyy="4.170414998500001e-05" iyz="0.0" izz="4.2604149625000006e-05"/>
    </inertial>
    <visual>
      <geometry>
        <mesh filename="package://rotors_description/meshes/propeller_cw.dae" scale="0.1 0.1 0.1"/>
        <!-- The propeller meshes have a radius of 1m -->
        <!-- <box size="${2*radius_rotor} 0.01 0.005"/> -->
      </geometry>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.005" radius="0.1"/>
        <!-- [m] -->
      </geometry>
    </collision>
  </link>
  <gazebo>
    <plugin filename="librotors_gazebo_motor_model.so" name="licasa1_back_right_motor_model">
      <robotNamespace>licasa1</robotNamespace>
      <jointName>rotor_3_joint</jointName>
      <linkName>rotor_3</linkName>
      <turningDirection>cw</turningDirection>
      <timeConstantUp>0.0125</timeConstantUp>
      <timeConstantDown>0.025</timeConstantDown>
      <maxRotVelocity>838000</maxRotVelocity>
      <motorConstant>8.54858e-06</motorConstant>
      <momentConstant>0.016</momentConstant>
      <commandSubTopic>gazebo/command/motor_speed</commandSubTopic>
      <motorNumber>3</motorNumber>
      <rotorDragCoefficient>8.06428e-05</rotorDragCoefficient>
      <rollingMomentCoefficient>1e-06</rollingMomentCoefficient>
      <motorSpeedPubTopic>motor_speed/3</motorSpeedPubTopic>
      <rotorVelocitySlowdownSim>10</rotorVelocitySlowdownSim>
    </plugin>
  </gazebo>
  <gazebo reference="rotor_3">
    <material>Gazebo/Blue</material>
  </gazebo>
  <joint name="rotor_4_joint" type="continuous">
    <origin rpy="0 0 0" xyz="0 -0.215 0.037"/>
    <axis xyz="0 0 1"/>
    <!-- TODO(ff): not currently set because it's not yet supported -->
    <!-- <limit effort="2000" velocity="${max_rot_velocity}" /> -->
    <parent link="base_link"/>
    <child link="rotor_4"/>
  </joint>
  <link name="rotor_4">
    <inertial>
      <mass value="0.005"/>
      <!-- [kg] -->
      <inertia ixx="9.7499961e-07" ixy="0.0" ixz="0.0" iyy="4.170414998500001e-05" iyz="0.0" izz="4.2604149625000006e-05"/>
    </inertial>
    <visual>
      <geometry>
        <mesh filename="package://rotors_description/meshes/propeller_ccw.dae" scale="0.1 0.1 0.1"/>
        <!-- The propeller meshes have a radius of 1m -->
        <!-- <box size="${2*radius_rotor} 0.01 0.005"/> -->
      </geometry>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.005" radius="0.1"/>
        <!-- [m] -->
      </geometry>
    </collision>
  </link>
  <gazebo>
    <plugin filename="librotors_gazebo_motor_model.so" name="licasa1_right_motor_model">
      <robotNamespace>licasa1</robotNamespace>
      <jointName>rotor_4_joint</jointName>
      <linkName>rotor_4</linkName>
      <turningDirection>ccw</turningDirection>
      <timeConstantUp>0.0125</timeConstantUp>
      <timeConstantDown>0.025</timeConstantDown>
      <maxRotVelocity>838000</maxRotVelocity>
      <motorConstant>8.54858e-06</motorConstant>
      <momentConstant>0.016</momentConstant>
      <commandSubTopic>gazebo/command/motor_speed</commandSubTopic>
      <motorNumber>4</motorNumber>
      <rotorDragCoefficient>8.06428e-05</rotorDragCoefficient>
      <rollingMomentCoefficient>1e-06</rollingMomentCoefficient>
      <motorSpeedPubTopic>motor_speed/4</motorSpeedPubTopic>
      <rotorVelocitySlowdownSim>10</rotorVelocitySlowdownSim>
    </plugin>
  </gazebo>
  <gazebo reference="rotor_4">
    <material>Gazebo/Blue</material>
  </gazebo>
  <joint name="rotor_5_joint" type="continuous">
    <origin rpy="0 0 0" xyz="0.18619546181355998 -0.1075 0.037"/>
    <axis xyz="0 0 1"/>
    <!-- TODO(ff): not currently set because it's not yet supported -->
    <!-- <limit effort="2000" velocity="${max_rot_velocity}" /> -->
    <parent link="base_link"/>
    <child link="rotor_5"/>
  </joint>
  <link name="rotor_5">
    <inertial>
      <mass value="0.005"/>
      <!-- [kg] -->
      <inertia ixx="9.7499961e-07" ixy="0.0" ixz="0.0" iyy="4.170414998500001e-05" iyz="0.0" izz="4.2604149625000006e-05"/>
    </inertial>
    <visual>
      <geometry>
        <mesh filename="package://rotors_description/meshes/propeller_cw.dae" scale="0.1 0.1 0.1"/>
        <!-- The propeller meshes have a radius of 1m -->
        <!-- <box size="${2*radius_rotor} 0.01 0.005"/> -->
      </geometry>
    </visual>
    <collision>
      <geometry>
        <cylinder length="0.005" radius="0.1"/>
        <!-- [m] -->
      </geometry>
    </collision>
  </link>
  <gazebo>
    <plugin filename="librotors_gazebo_motor_model.so" name="licasa1_front_right_motor_model">
      <robotNamespace>licasa1</robotNamespace>
      <jointName>rotor_5_joint</jointName>
      <linkName>rotor_5</linkName>
      <turningDirection>cw</turningDirection>
      <timeConstantUp>0.0125</timeConstantUp>
      <timeConstantDown>0.025</timeConstantDown>
      <maxRotVelocity>838000</maxRotVelocity>
      <motorConstant>8.54858e-06</motorConstant>
      <momentConstant>0.016</momentConstant>
      <commandSubTopic>gazebo/command/motor_speed</commandSubTopic>
      <motorNumber>5</motorNumber>
      <rotorDragCoefficient>8.06428e-05</rotorDragCoefficient>
      <rollingMomentCoefficient>1e-06</rollingMomentCoefficient>
      <motorSpeedPubTopic>motor_speed/5</motorSpeedPubTopic>
      <rotorVelocitySlowdownSim>10</rotorVelocitySlowdownSim>
    </plugin>
  </gazebo>
  <gazebo reference="rotor_5">
    <material>Gazebo/Red</material>
  </gazebo>

<!-- <link name="world"/> -->

<!-- <link name="firefly/base_link"/> -->

<!-- <xacro:include filename="$(find rotors_description)/urdf/firefly.xacro" /> -->

<!-- TESTBENCH 
<joint name="base_joint" type="fixed">
  <origin xyz="0 0 1.85"/>
  <parent link="world"/>
  <child link="testbench"/> 
</joint> -->

<!-- TESTBENCH -->
<joint name="fixed_joint" type="fixed">
  <origin xyz="0 0 2"/>
  <parent link="base_link"/>
  <child link="testbench"/> 
</joint> 

<link name="testbench">
  <inertial>
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <mass value="10" />
    <inertia ixx="1e-03" ixy="0.0" ixz="0.0" iyy="1e-03" iyz="0.0" izz="1e-03"/>
  </inertial>
  <visual> 
    <!-- <origin rpy="0 0 0" xyz="0 0 0"/>
    <geometry>
      <box size="0.15 0.15 0.01"/>
    </geometry> -->
    <origin xyz="0 0 0"/>
    <geometry>
      <mesh filename="package://uav-cable-suspended-robots-ros-pkg/mesh/licas/Assembly_X4.stl" scale=".001 .001 .001"/>
    </geometry>
  </visual>
</link> 

<!-- CABLES -->
<joint name="revolute_joint_z" type="revolute" >
  <parent link="testbench"/>
  <child link="bar_link_z" />
  <axis xyz="0 0 1" />
  <dynamics damping="0.517" friction="0.0"/>
  <limit effort="10" lower="${-360/180*pi}" upper="${360/180*pi}" velocity="6"/>
</joint>

<link name="bar_link_z">
  <inertial>
    <origin rpy="0 0 0" xyz="0 0 -0.4"/>
    <mass value="1"/>
    <inertia ixx="0.313" ixy="0.0" ixz="0.0" iyy="0.363" iyz="0.0" izz="0.745"/>
  </inertial>
</link>

<joint name="revolute_joint_x" type="revolute" >
  <parent link="bar_link_z" />
  <child link="bar_link_x" />
  <axis xyz="1 0 0" />
  <dynamics damping="1.33" friction="0.0"/>
  <limit effort="10.0" lower="${-150/180*pi}" upper="${150/180*pi}" velocity="6"/>    
</joint>

<link name="bar_link_x">
  <inertial>
    <origin rpy="0 0 0" xyz="0 0 -0.174"/>
    <mass value="4.237"/>
    <inertia ixx="0.313" ixy="0.0" ixz="0.0" iyy="0.363" iyz="0.0" izz="0.745"/>
  </inertial>
</link>

<joint name="revolute_joint_y" type="revolute" >
  <parent link="bar_link_x" />
  <child link="bar_link_y" />
  <axis xyz="0 1 0" />
  <dynamics damping="1.311" friction="0.0"/>
  <limit effort="10.0" lower="${-150/180*pi}" upper="${150/180*pi}" velocity="6"/>
</joint>

<link name="bar_link_y">
  <inertial>
    <origin rpy="0 0 0" xyz="0 0 -0.824"/>
    <mass value="5.032"/>
    <inertia ixx="0.313" ixy="0.0" ixz="0.0" iyy="0.363" iyz="0.0" izz="0.745"/>
  </inertial>
  <visual>
    <origin rpy="0 0 0" xyz="0 0 -0.5"/>
    <geometry>
      <cylinder length="1.0" radius="0.01"/>
    </geometry>
  </visual>
</link>


<!-- SHOULDER -->
<joint name="shoulder_joint_z" type="revolute" >
  <origin xyz="0 0 -1.0"/>
  <parent link="bar_link_y"/>
  <child link="shoulder_link_z"/>
  <axis xyz="0 0 1"/>
  <dynamics damping="0.0" friction="0.0"/>
  <limit effort="6.0" lower="${-150/180*pi}" upper="${150/180*pi}" velocity="6"/>
</joint>

<link name="shoulder_link_z">
  <inertial>
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <mass value="0.639"/>
    <inertia ixx="3.02e-3" ixy="0" ixz="0" iyy="8.64e-4" iyz="0" izz="3.39e-3"/>
  </inertial>
</link>

<joint name="shoulder_joint_x" type="revolute" >
  <origin xyz="0 0 0"/>
  <parent link="shoulder_link_z"/>
  <child link="shoulder_link_x"/>
  <axis xyz="1 0 0"/>
  <dynamics damping="0.0" friction="0.0"/>
  <limit effort="6.0" lower="${-150/180*pi}" upper="${150/180*pi}" velocity="6"/>
</joint>

<link name="shoulder_link_x">
  <inertial>
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <mass value="0.639"/>
    <inertia ixx="3.02e-3" ixy="0" ixz="0" iyy="8.64e-4" iyz="0" izz="3.39e-3"/>
  </inertial>
</link>


<joint name="shoulder_joint_y" type="revolute" >
  <origin xyz="0 0 0"/>
  <parent link="shoulder_link_x"/>
  <child link="shoulder_link_y"/>
  <axis xyz="0 1 0"/>
  <dynamics damping="0.0" friction="0.0"/>
  <limit effort="6.0" lower="${-150/180*pi}" upper="${150/180*pi}" velocity="6"/>
</joint>

<link name="shoulder_link_y">
  <inertial>
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <mass value="0.639"/>
    <inertia ixx="3.02e-3" ixy="0" ixz="0" iyy="8.64e-4" iyz="0" izz="3.39e-3"/>
  </inertial>
  <visual>
    <origin xyz="0 0 0"/>
    <geometry>
      <mesh filename="package://uav-cable-suspended-robots-ros-pkg/mesh/licas/Shoulder_Pitch_Structure.dae" scale=".001 .001 .001"/>
    </geometry>
  </visual>
  <collision>
    <origin xyz="0 0 0"/>
    <geometry>
      <mesh filename="package://uav-cable-suspended-robots-ros-pkg/mesh/licas/Shoulder_Pitch_Structure.dae" scale=".001 .001 .001"/>
    </geometry>
  </collision>
</link>


<!-- LEFT ARM-->
<joint name="LiCAS_A1_q1_1" type="revolute">
  <origin xyz="0 0.14 0.0"/>
  <parent link="shoulder_link_y"/>
  <child link="left_shoulder_roll"/>
  <axis xyz="0 1 0"/>
  <dynamics damping="0.0" friction="0.0"/>
  <limit effort="10.0" lower="${-90/180*pi}" upper="${90/180*pi}" velocity="6"/>
</joint> 

<link name="left_shoulder_roll">
  <inertial>
    <origin xyz="0 0.0236 -0.00946"/> <!-- <origin xyz="0 0.0164 -0.00946"/> -->
    <mass value="0.233"/>
    <inertia ixx="3.68e-4" ixy="-7.57e-6" ixz="5.38e-6" iyy="1.16e-4" iyz="-4.8e-5" izz="3.76e-4"/>
  </inertial>
  <visual>
    <origin xyz="0 0.04 0"/>
    <geometry>
      <mesh filename="package://uav-cable-suspended-robots-ros-pkg/mesh/licas/Shoulder_Roll_Structure_Left.dae" scale=".001 .001 .001"/>
    </geometry>
  </visual>
  <collision>
    <origin xyz="0 0.04 0"/>
    <geometry>
      <mesh filename="package://uav-cable-suspended-robots-ros-pkg/mesh/licas/Shoulder_Roll_Structure_Left.dae" scale=".001 .001 .001"/>
    </geometry>
  </collision>
</link>


<joint name="LiCAS_A1_q1_2" type="revolute">
  <origin xyz="0 0.04 0"/>
  <parent link="left_shoulder_roll"/>
  <child link="left_shoulder_yaw"/>
  <axis xyz="1 0 0"/>
  <dynamics damping="0.0" friction="0.0"/>
  <limit effort="10.0" lower="${-20/180*pi}" upper="${90/180*pi}" velocity="6"/>
</joint>

<link name="left_shoulder_yaw">
  <inertial>
    <origin xyz="-0.015 0 -0.1"/>
    <mass value="0.246"/>
    <inertia ixx="3.49e-4" ixy="0" ixz="-3.79e-5" iyy="4.46e-4" iyz="0" izz="1.49e-4"/>
  </inertial>
  <visual>
    <origin xyz="0 0 0"/>
    <geometry>
      <mesh filename="package://uav-cable-suspended-robots-ros-pkg/mesh/licas/Assembly_Shoulder_Yaw_Structure.dae" scale=".001 .001 .001"/>
    </geometry>
  </visual>
  <collision>
    <origin xyz="0 0 0"/>
    <geometry>
      <mesh filename="package://uav-cable-suspended-robots-ros-pkg/mesh/licas/Assembly_Shoulder_Yaw_Structure.dae" scale=".001 .001 .001"/>
    </geometry>
  </collision>
</link> 


<joint name="LiCAS_A1_q1_3" type="revolute">
  <origin xyz="0 0 -0.143"/>
  <parent link="left_shoulder_yaw"/>
  <child link="left_elbow_pitch"/>
  <axis xyz="0 0 1"/>
  <dynamics damping="0.0" friction="0.0"/>
  <limit effort="10.0" lower="${-90/180*pi}" upper="${90/180*pi}" velocity="6"/>
</joint>

<link name="left_elbow_pitch">
  <inertial>
    <origin xyz="0 0 -0.093"/>
    <mass value="0.214"/>
    <inertia ixx="4.21e-4" ixy="-7.81e-7" ixz="-4.06e-5" iyy="4.26e-4" iyz="1.8e-5" izz="5.01e-5"/>
  </inertial>
  <visual>
    <origin xyz="0 0 -0.132"/>
    <geometry>
      <mesh filename="package://uav-cable-suspended-robots-ros-pkg/mesh/licas/Elbow_Pitch_Structure_Left_Cut.dae" scale=".001 .001 .001"/>
    </geometry>
  </visual>
  <collision>
    <origin xyz="0 0 -0.132"/>
    <geometry>
      <mesh filename="package://uav-cable-suspended-robots-ros-pkg/mesh/licas/Elbow_Pitch_Structure_Left_Cut.dae" scale=".001 .001 .001"/>
    </geometry>
  </collision>
</link>


<joint name="LiCAS_A1_q1_4" type="revolute">
  <origin xyz="0 0 -0.132"/>  
  <parent link="left_elbow_pitch"/>
  <child link="left_forearm_link"/>
  <axis xyz="0 1 0"/>
  <dynamics damping="0.0" friction="0.0"/>
  <limit effort="10.0" lower="${-150/180*pi}" upper="${150/180*pi}" velocity="6"/>
</joint>

<link name="left_forearm_link">
  <inertial>
    <origin xyz="0 0 -0.092"/>
    <mass value="0.106"/>
    <inertia ixx="3.94e-4" ixy="0" ixz="0" iyy="3.75e-4" iyz="-3.5e-6" izz="3.05e-5"/>
  </inertial>
  <visual>
    <origin xyz="0 0 0"/>
    <geometry>
      <mesh filename="package://uav-cable-suspended-robots-ros-pkg/mesh/licas/Forearm_Link_with_Gripper_Servo.dae" scale=".001 .001 .001"/>
    </geometry>
  </visual>
  <collision>
    <origin xyz="0 0 0"/>
    <geometry>
      <mesh filename="package://uav-cable-suspended-robots-ros-pkg/mesh/licas/Forearm_Link_with_Gripper_Servo.dae" scale=".001 .001 .001"/>
    </geometry>
  </collision>
</link> 


<!-- RIGHT ARM-->
<joint name="LiCAS_A1_q2_1" type="revolute">
  <origin xyz="0 -0.14 0.0"/>
  <parent link="shoulder_link_y"/>
  <child link="right_shoulder_roll"/>
  <axis xyz="0 1 0"/>
  <dynamics damping="0.0" friction="0.0"/>
  <limit effort="10.0" lower="${-90/180*pi}" upper="${90/180*pi}" velocity="6"/>
</joint>

<link name="right_shoulder_roll">
  <inertial>
    <origin xyz="0 -0.0236 -0.00946"/> <!-- <origin xyz="0 -0.0164 -0.00946"/> -->
    <mass value="0.233"/>
    <inertia ixx="3.68e-4" ixy="7.57e-6" ixz="5.38e-6" iyy="1.16e-4" iyz="4.8e-5" izz="3.76e-4"/>
  </inertial>
  <visual>
    <origin xyz="0 -0.04 0"/>
    <geometry>
      <mesh filename="package://uav-cable-suspended-robots-ros-pkg/mesh/licas/Shoulder_Roll_Structure_Right.dae" scale=".001 .001 .001"/>
    </geometry>
  </visual>
  <collision>
    <origin xyz="0 -0.04 0"/>
    <geometry>
      <mesh filename="package://uav-cable-suspended-robots-ros-pkg/mesh/licas/Shoulder_Roll_Structure_Right.dae" scale=".001 .001 .001"/>
    </geometry>
  </collision>
</link>


<joint name="LiCAS_A1_q2_2" type="revolute">
  <origin xyz="0 -0.04 0"/>
  <parent link="right_shoulder_roll"/>
  <child link="right_shoulder_yaw"/>
  <axis xyz="1 0 0"/>
  <dynamics damping="0.0" friction="0.0"/>
  <limit effort="10.0" lower="${-90/180*pi}" upper="${20/180*pi}" velocity="6"/>
</joint>

<link name="right_shoulder_yaw">
  <inertial>
    <origin xyz="-0.015 0 -0.1"/>
    <mass value="0.246"/>
    <inertia ixx="3.49e-4" ixy="0" ixz="-3.79e-5" iyy="4.46e-4" iyz="0" izz="1.49e-4"/>
  </inertial>
  <visual>
    <origin xyz="0 0 0"/>
    <geometry>
      <mesh filename="package://uav-cable-suspended-robots-ros-pkg/mesh/licas/Assembly_Shoulder_Yaw_Structure.dae" scale=".001 .001 .001"/>
    </geometry>
  </visual>
  <collision>
    <origin xyz="0 0 0"/>
    <geometry>
      <mesh filename="package://uav-cable-suspended-robots-ros-pkg/mesh/licas/Assembly_Shoulder_Yaw_Structure.dae" scale=".001 .001 .001"/>
    </geometry>
  </collision>
</link> 


<joint name="LiCAS_A1_q2_3" type="revolute">
  <origin xyz="0 0 -0.143"/>  
  <parent link="right_shoulder_yaw"/>
  <child link="right_elbow_pitch"/>
  <axis xyz="0 0 1"/>
  <dynamics damping="0.0" friction="0.0"/>
  <limit effort="10.0" lower="${-90/180*pi}" upper="${90/180*pi}" velocity="6"/>
</joint> 

<link name="right_elbow_pitch">
  <inertial>
    <origin xyz="0 0 -0.093"/>  
    <mass value="0.214"/>
    <inertia ixx="4.21e-4" ixy="7.81e-7" ixz="-4.06e-5" iyy="4.26e-4" iyz="-1.8e-5" izz="5.01e-5"/>
  </inertial>
  <visual>
    <origin xyz="0 0 -0.132"/> 
    <geometry>
      <mesh filename="package://uav-cable-suspended-robots-ros-pkg/mesh/licas/Elbow_Pitch_Structure_Right_Cut.dae" scale=".001 .001 .001"/>
    </geometry>
  </visual>
  <collision>
    <origin xyz="0 0 -0.132"/>
    <geometry>
      <mesh filename="package://uav-cable-suspended-robots-ros-pkg/mesh/licas/Elbow_Pitch_Structure_Right_Cut.dae" scale=".001 .001 .001"/>
    </geometry>
  </collision>
</link>


<joint name="LiCAS_A1_q2_4" type="revolute">
  <origin xyz="0 0 -0.132"/>  
  <parent link="right_elbow_pitch"/>
  <child link="right_forearm_link"/>
  <axis xyz="0 1 0"/>
  <dynamics damping="0.0" friction="0.0"/>
  <limit effort="10.0" lower="${-150/180*pi}" upper="${150/180*pi}" velocity="6"/>
</joint>

<link name="right_forearm_link">
  <inertial>
    <origin xyz="0 0 -0.092"/>
    <mass value="0.106"/>
    <inertia ixx="3.94e-4" ixy="0" ixz="0" iyy="3.75e-4" iyz="3.5e-6" izz="3.05e-5"/>
  </inertial>
  <visual>
    <origin xyz="0 0 0"/> 
    <geometry>
      <mesh filename="package://uav-cable-suspended-robots-ros-pkg/mesh/licas/Forearm_Link_with_Gripper_Servo.dae" scale=".001 .001 .001"/>
    </geometry>
  </visual>
  <collision>
    <origin xyz="0 0 0"/>
    <geometry>
      <mesh filename="package://uav-cable-suspended-robots-ros-pkg/mesh/licas/Forearm_Link_with_Gripper_Servo.dae" scale=".001 .001 .001"/>
    </geometry>
  </collision>
</link>

<!-- END EFFECTORS -->

<joint name="right_eef_joint" type="fixed">
  <origin xyz="0 0 -0.25"/>
  <parent link="right_forearm_link"/>
  <child link="right_eef_link"/>
</joint>

<link name="right_eef_link">
  <inertial>
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <mass value="1e-3" />
    <inertia ixx="1e-03" ixy="0.0" ixz="0.0" iyy="1e-03" iyz="0.0" izz="1e-03"/>
  </inertial>
</link>

<joint name="left_eef_joint" type="fixed">
  <origin xyz="0 0 -0.25"/>
  <parent link="left_forearm_link"/>
  <child link="left_eef_link"/>
</joint>

<link name="left_eef_link">
  <inertial>
    <origin rpy="0 0 0" xyz="0 0 0"/>
    <mass value="1e-3" />
    <inertia ixx="1e-03" ixy="0.0" ixz="0.0" iyy="1e-03" iyz="0.0" izz="1e-03"/>
  </inertial>
</link> 

</robot>
